### PR TITLE
refactor(stream): introduce `Successor` to refine `SortBuffer`

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

1. introduce `Successor` trait
2. refactor `SortBuffer` to use `Successor`

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/stream/src/executor/sort_buffer.rs [hunk 1]
```diff
@@ -140,36 +140,25 @@ impl<S: StateStore> SortBuffer<S> {
...
-                // should be output, but it's hard to represent `OwnedRow::MAX`. A possible
-                // implementation is introducing `next_unit` on a subset of `ScalarImpl` variants.
-                // Currently, we can skip some values explicitly.
-                Bound::Excluded((last_watermark, OwnedRow::empty().into()))
+                Bound::Excluded((
+                    // TODO: unsupported type or watermark overflow. Do we have better ways instead
+                    // of unwrap?
+                    last_watermark.successor().unwrap(),
+                    OwnedRow::empty().into(),
+                ))
```
Comment (st1page) [inline src/stream/src/executor/sort_buffer.rs:148]: I think it is ok and the optimizer should ensure that. Also, maybe a better error message here is better.

**Review by st1page — APPROVED** [review]
LGTM

File: src/stream/src/executor/sort_buffer.rs (hunk 1 above)
Comment (BugenZhao) [reply src/stream/src/executor/sort_buffer.rs:148]: Or do we need to use `Unbounded` when overflowing (corner case, never mind 😄)?

**Review by BugenZhao — APPROVED** [review]

File: src/stream/src/executor/sort_buffer.rs (hunk 1 above)
Comment (TennyZhuang) [reply src/stream/src/executor/sort_buffer.rs:148]: We can discuss them later

**Pull request merged**